### ISSUE TITLE
module loading error when used with webpack2 CommonsChunkPlugin

### ISSUE THREAD

**Issue by jbellenger** [rxaviers/globalize-webpack-plugin#44]
In some cases, it looks like using [CommonsChunkPlugin](https://webpack.js.org/plugins/commons-chunk-plugin/) with globalize-webpack-plugin can break module loading. This appears to happen when using a `CommonsChunkPlugin.maxChunks` value that is greater or equal to the number of supported locales. The same error does not happen when using webpack1. This may be related to #38, though after hacking up the regex described in #38 I've found the error is still reproducible.

I've put together a small test case at [jbellenger/globalize-webpack-plugin-testcase](https://github.com/jbellenger/globalize-webpack-plugin-testcase).

I'm open to putting together a PR, though I'm presently at a loss for why this is happening and could use some pointers on where to look.

**Comment by rxaviers** [comment]
Thanks, will look into it.

**Comment by rxaviers** [comment]
@jbellenger thanks for your detailed test case and sorry for my delay on finding a time to look into this.

Globalize plugin [dynamically creates a chunk per supported locale](https://github.com/rxaviers/globalize-webpack-plugin/blob/master/ProductionModePlugin.js#L148) in the build (production) phase. Did you check what's the relationship between CommonChunkPlugin `options.minChunks` and that?

**Comment by jbellenger** [comment]
Thanks for the pointer!
I experimented with disabling the section that moved modules into the compiledDataChunks. The result of disabling chunk moving was that the 'bad' build in my demo project logs only 1 error instead of 2:
- the en bundle still fails to load some unknown module
- the main bundle is now able to load Globalize, which it couldn't do with chunk moving.

I've looked into CommonsChunkPlugin and ProductionModePlugin, but haven't been able to identify the cause. It's also possible I made an error in applying the webpack2 migration process to [the globalize-webpack example project](https://github.com/globalizejs/globalize/blob/master/examples/app-npm-webpack/webpack-config.js) to produce [my demo webpack config](https://github.com/jbellenger/globalize-webpack-plugin-testcase/blob/master/webpack-config.js)

**Comment by necolas** [comment]
@rxaviers do you need help from the webpack team here (cc @TheLarkInn)? This has been blocking our upgrade to webpack@2 – or using webpack@1 with deterministic hashing – which we know will significantly improve load times following deploys.

(I'm also curious why the compiler doesn't pull the strings out at build time, hash them, and replace the string with a hash that can be used to directly look up the string in the map.)

**Comment by rxaviers** [comment]
@necolas, a help from the webpack team would be definitely welcome. Thanks

> (I'm also curious why the compiler doesn't pull the strings out at build time, hash them, and replace the string with a hash that can be used to directly look up the string in the map.)

Are you referring to hashing the message keys or https://github.com/rxaviers/react-globalize-webpack-plugin/issues/21?

**Comment by necolas** [comment]
@TheLarkInn we want to get onto webpack@2 so we can benefit from all the improvements it brings. However, this library is currently an important part of our build and needs changes to work with stable modules id's. It looks like a key part of the problem lies in generating dynamic chunks and using CommonsChunkPlugin + minChunks. Would appreciate any time you have to give us pointers or insights that can help us determine the way forward.

**Comment by TheLarkInn** [comment]
I'll take some time to look this over. Is there a way you could isolate the error within a tiny repo? (I will attempt to do so myself as well) if so I can then find out what's causing the problem.

**Comment by TheLarkInn** [comment]
Ah sorry I read original issue where there is a repro. I'll look into this tonight!

**Comment by necolas** [comment]
Thanks so much!

**Comment by jbellenger** [comment]
@TheLarkInn : were you able to look into this? Any insight as to what the issue is?

**Comment by TheLarkInn** [comment]
I've been getting a lot of push with work (including late nights), but I managed to reproduce the error at the least and have not got past that though. I'm hoping to have at least some sort of direction by Friday at the latest.

**Comment by jbellenger** [comment]
I took another look into this this week, and think that I am closer to understanding where the issue might be.

The most refined description of the issue that I can muster:
> when globalize-webpack-plugin is combined with an async CommonsChunkPlugin, the compilation produces a locale chunk that calls `webpackJsonp` with what appears to be an invalid value for `executeModules`. The values in executeModules are defined in the async chunk, though the module bodies themselves are empty and the async chunk isn't loaded by the runtime.

Example build results: https://gist.github.com/jbellenger/2eac00cc0179254bde92bcbb4b7dba7b

I _suspect_ that the problem may be related to the way that globalize-webpack-plugin handles the [entry-option hook](https://github.com/rxaviers/globalize-webpack-plugin/blob/master/ProductionModePlugin.js#L110):
```
  compiler.plugin("entry-option", function(context) {
    supportedLocales.forEach(function(locale) {
      compiler.apply(new MultiEntryPlugin(context, [], "globalize-compiled-data-" + locale ));
    });
  });
```
The MultiEntryPlugin's created here go through [Compilation.addEntry](https://github.com/webpack/webpack/blob/master/lib/Compilation.js#L458) and [Compilation._addModuleChain](https://github.com/webpack/webpack/blob/master/lib/Compilation.js#L362). Suspiciously, because the MultiEntryPlugin's created do not specify dependencies, they appear to create a module that is reused across multiple locale chunks. 

Screenshot showing that inserting a new entry for the en chunk results in reusing the module created for the ar language chunk:
![screen shot 2017-04-27 at 5 09 31 pm](https://cloud.githubusercontent.com/assets/1391805/25509300/bbe5905c-2b6c-11e7-9aee-75a56c1e1e98.png)

@TheLarkInn: I could use some expert webpack guidance on if this is usage of MultiEntryPlugin is correct for webpack2, and if some of the observed results are as expected.

**Comment by jbellenger** [comment]
After some more investigation, I've opened [webpack 4795](https://github.com/webpack/webpack/issues/4795)

**Comment by rxaviers** [comment]
@jbellenger, thanks for the investigation and for keeping us posted.

**Comment by jbellenger** [comment]
I simplified my testcase to remove globalize-webpack-plugin entirely, and can confirm that the issue still happens. So it's pretty likely that this is a pure webpack issue, though one that globalize-webpack-plugin users are likely to run into when using webpack 2. 

I'm closing this issue. Follow-up will happen in [webpack 4795](https://github.com/webpack/webpack/issues/4795)

**Comment by TheLarkInn** [comment]
Apologize for the delay on following up with this. Extremely relieved that @jbellenger isolated and discovered the source of the issue. I talked with Sokra about this and have prioritized it to critical (top of our todo).

**Comment by rxaviers** [comment]
Aweome!

**Comment by TheLarkInn** [comment]
We fixed related bug in https://github.com/webpack/webpack/pull/4815 

If possible we would love for you test and confirm against that commithash.

**Comment by jbellenger** [comment]
Thanks @TheLarkInn.
Confirmed that this is fixed with webpack 2.5.0
